### PR TITLE
Python: Fix conversation-id propagation when chat_options is a dict

### LINKED ISSUE #4305: Python: Function invocation conversation-id propagation crashes when chat_options in kwargs is a dict
### Description
Function invocation conversation-id propagation crashes when `chat_options` in kwargs is a dict.

`_update_conversation_id` does:
```python
kwargs["chat_options"].conversation_id = conversation_id
```
which assumes an object with attribute access. If a dict is provided, this raises `AttributeError`.

### Code Sample
```python
from agent_framework._tools import _update_conversation_id

kwargs = {"chat_options": {}}
_update_conversation_id(kwargs, "conv_1", {})
# Actual: AttributeError: 'dict' object has no attribute 'conversation_id'
# Expected: kwargs['chat_options']['conversation_id'] updated or graceful handling
```

A full invocation-loop repro also fails in the same spot when tool-calling updates conversation id.

### Error Messages / Stack Traces
```text
AttributeError: 'dict' object has no attribute 'conversation_id' and no __dict__ for setting new attributes
```

### Package Versions
agent-framework-core: source checkout at commit `02ba27493`

### Python Version
Python 3.13.5

### Additional Context
This is in a hot path used by the function invocation loop for conversation-based providers.

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

When `chat_options` was passed as a plain `dict` (or `TypedDict`) in kwargs, `_update_conversation_id` crashed with an `AttributeError` because it unconditionally used attribute assignment (`chat_opts.conversation_id = ...`) instead of key-based access.

Fixes #4305

### Description

The root cause was that `_update_conversation_id` assumed `chat_options` was always an object supporting attribute assignment, but callers could pass it as a `dict` or `TypedDict`. The fix adds an `isinstance(chat_opts, dict)` check and uses dictionary key assignment for dict-like values while preserving attribute assignment for object-style options. Unit tests covering dict, TypedDict, object, fallback, and no-op cases were added to prevent regression.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by moonbox3's agent

### CODE REVIEW

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 3 | **Confidence:** 92%

### ✓ Correctness
> The change correctly extends `_update_conversation_id` to handle `chat_options` passed as a plain dict (or TypedDict) in addition to an object with attribute access. The `isinstance(chat_opts, dict)` check is correct since TypedDict instances are dicts at runtime, and the fallback to attribute-based access for non-dict objects preserves existing behavior. Tests cover the key scenarios adequately. No correctness issues found.

### ✓ Security Reliability
> This diff adds a type check so that `_update_conversation_id` handles `chat_options` as either a dict (key access) or an object (attribute access), along with thorough unit tests. The change is straightforward and low-risk from a security and reliability perspective. No injection risks, secrets, resource leaks, or unsafe deserialization are introduced. The isinstance check is a reasonable way to distinguish the two cases.

### ✓ Test Coverage
> The new tests in TestUpdateConversationId provide solid coverage of the dict-vs-object branching logic added to _update_conversation_id. All key code paths are exercised: dict chat_options, TypedDict chat_options, object chat_options, absent chat_options, None conversation_id early return, and the optional options parameter. Assertions are meaningful and verify correct values rather than just absence of exceptions. No blocking issues found.

### Suggestions
- Minor: consider whether a Mapping ABC check (`isinstance(chat_opts, collections.abc.Mapping)`) would be more robust than `isinstance(chat_opts, dict)`, in case a non-dict mapping type is ever passed. Not blocking since current callers only use dict/TypedDict/objects.
- Consider whether a `TypedDict` subclass could bypass the `isinstance(chat_opts, dict)` check in edge cases (it shouldn't, since `TypedDict` instances are plain dicts at runtime, but worth a mental note).
- Consider adding a test that passes an object-style chat_options together with the optional `options` dict to verify both are updated correctly in the attribute-access branch (currently `test_options_dict_also_updated` only exercises the dict branch).
- Consider a test where the dict already contains a previous conversation_id to verify it gets overwritten, confirming idempotent behavior.

---
*Automated review by moonbox3's agents*

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py">_tools.py</a></td><td>783</td><td>80</td><td>89%</td><td><a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L168-L169">168&ndash;169</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L328">328</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L330">330</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L348-L350">348&ndash;350</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L358">358</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L376">376</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L390">390</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L397">397</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L404">404</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L420">420</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L422">422</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L429">429</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L437">437</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L469">469</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L494">494</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L498">498</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L515-L517">515&ndash;517</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L564-L566">564&ndash;566</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L588">588</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L644">644</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L666">666</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L733-L739">733&ndash;739</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L775">775</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L786-L797">786&ndash;797</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L816-L818">816&ndash;818</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L822">822</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L826">826</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L840-L842">840&ndash;842</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1181">1181</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1201">1201</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1277-L1281">1277&ndash;1281</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1405">1405</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1409">1409</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1433">1433</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1551">1551</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1581">1581</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1601">1601</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1603">1603</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1656">1656</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1719">1719</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1910-L1911">1910&ndash;1911</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L1962">1962</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L2031-L2032">2031&ndash;2032</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L2092">2092</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L2097">2097</a>, <a href="https://github.com/microsoft/agent-framework/blob/f17c735c405c8698099d83ba1af28189fa1bef7a/packages/core/agent_framework/_tools.py#L2104">2104</a></td></tr><tr><td><b>TOTAL</b></td><td><b>22681</b></td><td><b>2564</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4938 | 20 :zzz: | 0 :x: | 0 :fire: | 1m 21s :stopwatch: |

File: python/packages/core/agent_framework/_tools.py [hunk 1]
```diff
@@ -1669,7 +1669,11 @@ def _update_conversation_id(
     if conversation_id is None:
         return
     if "chat_options" in kwargs:
-        kwargs["chat_options"].conversation_id = conversation_id
+        chat_opts = kwargs["chat_options"]
+        if isinstance(chat_opts, dict):
```
Comment (eavanvalkenburg) [inline python/packages/core/agent_framework/_tools.py]: Chat options should always be a dict, afaik

**Review by giles17 — APPROVED** [review]

File: python/packages/core/agent_framework/_tools.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_tools.py]: Fixed in `8fde5fbb`.

Comment (moonbox3) [reply python/packages/core/agent_framework/_tools.py]: Fixed.
